### PR TITLE
SNOW-334890: External browser SSO authentication with proxy

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [ ] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [ ] Create tests which fail without the change (if possible)
- [ ] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [ ] Extend the README / documentation and ensure is properly displayed (if necessary)
- [ ] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/services/sso_url_provider.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        "ACCOUNT_NAME": account,
+        "LOGIN_NAME": user,
+        "PORT": port,
+        "PROTOCOL": protocol,
+        "AUTHENTICATOR": authenticator,
+        "BROWSER_MODE_REDIRECT_PORT": callback_port.toString()
+      }
+    };
+
+    const httpsClient = new HttpClient(connectionConfig)
```
Comment (sfc-gh-dprzybysz) [inline lib/services/sso_url_provider.js:59]: should we create new http client here or used a service?

File: lib/services/sso_url_provider.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    const agent = httpsClient.getAgent(url, connectionConfig.getProxy());
```
Comment (sfc-gh-dprzybysz) [inline lib/services/sso_url_provider.js:60]: to get an agent we have to use httpsClient? cannot we just use the same client?

File: lib/services/sso_url_provider.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        url: url,
+        headers: header,
+        data: body,
+        requestOCSP: false,
+        rejectUnauthorized: true,
+        httpsAgent: agent
+      };
+
+    // Post request to get the SSO URL
+    return axios.request(requestOptions)
```
Comment (sfc-gh-dprzybysz) [inline lib/services/sso_url_provider.js:74]: cannot we use here our http client?

File: test/unit/authentication/authentication_test.js [hunk 4]
```diff
@@ -86,41 +89,35 @@ describe('external browser authentication', function ()
...
-      {
-        var data =
-        {
-          data: {
-            data:
-            {
-              ssoUrl: mockSsoURL,
-              proofKey: mockProofKey
-            }
+    mock('ssoUrlProvider', {
```
Comment (sfc-gh-dprzybysz) [inline test/unit/authentication/authentication_test.js:92]: cannot we still mock at the axios level?

File: lib/services/sso_url_provider.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
+/*
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/services/sso_url_provider.js:1]: ok

File: lib/services/sso_url_provider.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+const Util = require('../util');
+const Errors = require('../errors');
+const HttpClient = require("../http/node");
+const axios = require("axios");
+const {rest} = require("../global_config");
+
+/**
+ * Creates a new instance of an LargeResultSetService.
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/services/sso_url_provider.js:12]: only this line :D. Fixed

File: lib/services/sso_url_provider.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @param {Object} connectionConfig
+ * @param {Object} httpClient
+ * @constructor
+ */
+function SsoUrlProvider(connectionConfig, httpClient) {
+  // validate input
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/services/sso_url_provider.js:19]: ok

File: lib/authentication/auth_web.js [hunk 8]
```diff
@@ -31,14 +31,10 @@ function auth_web(host, browserActionTimeout, webbrowser, httpclient, ) {
   }
 
   const open = typeof webbrowser !== "undefined" ? webbrowser : require('open');
-  const axios = typeof httpclient !== "undefined" ? httpclient : require('axios');
 
-  const browserTimeout = browserActionTimeout
-  const port = rest.HTTPS_PORT;
-  const protocol = rest.HTTPS_PROTOCOL;
+  let ssoURL;
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:35]: could be local

File: lib/services/sso_url_provider.js (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply lib/services/sso_url_provider.js]: we can but I prefer not use request method. I've add our client post directly using axios post method.

File: test/unit/authentication/authentication_test.js (hunk 4 above)
Comment (sfc-gh-pmotacki) [reply test/unit/authentication/authentication_test.js]: fixed

File: lib/authentication/auth_web.js [hunk 9]
```diff
@@ -3,25 +3,27 @@
  */
 
 const util = require('../util');
-const rest = require('../global_config').rest;
-
 const net = require('net');
 const querystring = require('querystring');
 const URLUtil = require('./../../lib/url_util');
 const Util = require('./../../lib/util');
+var SsoUrlProvider = require('../authentication/sso_url_provider');
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:10]: let it be const

File: lib/authentication/sso_url_provider.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    };
+
+    const agent = httpClient.getAgent(url, connectionConfig.getProxy());
+
+    const requestOptions =
+      {
+        headers: header,
+        requestOCSP: false,
+        rejectUnauthorized: true,
+        httpsAgent: agent
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/sso_url_provider.js:64]: do we need explicitly setting agent here? our http client should already be proxy aware?

File: lib/connection/connection.js [hunk 11]
```diff
@@ -236,7 +236,7 @@ function Connection(context)
     }
 
     // Get authenticator to use
-    var auth = Authenticator.getAuthenticator(connectionConfig);
+    var auth = Authenticator.getAuthenticator(connectionConfig, context.getServices().ssoUrlProvider);
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection.js:239]: let it be const

File: lib/connection/connection.js [hunk 12]
```diff
@@ -297,7 +297,7 @@ function Connection(context)
     var self = this;
 
     // Get authenticator to use
-    var auth = Authenticator.getAuthenticator(connectionConfig);
+    var auth = Authenticator.getAuthenticator(connectionConfig, context.getHttpClient());
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection.js:300]: const?

File: lib/http/base.js [hunk 13]
```diff
@@ -192,6 +192,8 @@ HttpClient.prototype.getAgent = function (url, proxy, mock) {
   return null;
 };
 
+HttpClient.prototype.post = axios.post;
```
Comment (sfc-gh-dprzybysz) [inline lib/http/base.js:195]: cannot we here hide all the agent configuration things?

File: lib/authentication/sso_url_provider.js (hunk 10 above)
Comment (sfc-gh-dprzybysz) [inline lib/authentication/sso_url_provider.js:64]: what about http agent? not only https we use

File: lib/services/sso_url_provider.js (hunk 2 above)
Comment (sfc-gh-dprzybysz) [reply lib/services/sso_url_provider.js]: but we need then to configure agents in two places - we can have it in one

File: lib/connection/connection.js (hunk 12 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection.js:300]: ok

File: lib/connection/connection.js (hunk 11 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection.js:239]: ok

File: lib/authentication/sso_url_provider.js (hunk 10 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/sso_url_provider.js:64]: but we have custom method to create agent `getAgent`. It contains all conditions like ocsp, proxy ,mock

File: lib/http/base.js [hunk 14]
```diff
@@ -192,8 +111,57 @@ HttpClient.prototype.getAgent = function (url, proxy, mock) {
...
+      data: data,
+      timeout: timeout,
+      requestOCSP: true,
+      httpsAgent: agent,
+      retryDelay: backoffStrategy,
+      rejectUnauthorized: true,
+      // we manually parse jsons or other structures from the server so they need to be text
+      responseType: options.responseType || 'text',
+    };
+  return requestOptions;
```
Comment (sfc-gh-dprzybysz) [inline lib/http/base.js:162]: can we inline this variable?

File: lib/http/base.js [hunk 15]
```diff
@@ -192,8 +111,57 @@ HttpClient.prototype.getAgent = function (url, proxy, mock) {
...
+
+  const requestOptions =
+    {
+      method: options.method,
+      url: options.url,
+      headers: headers,
+      data: data,
+      timeout: timeout,
+      requestOCSP: true,
+      httpsAgent: agent,
```
Comment (sfc-gh-dprzybysz) [inline lib/http/base.js:156]: should we also set httpAgent?

File: lib/authentication/sso_url_provider.js (hunk 10 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/sso_url_provider.js]: added

File: lib/http/base.js (hunk 15 above)
Comment (sfc-gh-pmotacki) [reply lib/http/base.js]: ok

File: lib/http/base.js (hunk 13 above)
Comment (sfc-gh-pmotacki) [reply lib/http/base.js:115]: it isn't needed now

File: lib/connection/connection.js (hunk 11 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection.js]: ok

File: lib/authentication/sso_url_provider.js (hunk 10 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/sso_url_provider.js]: It is set inside httpClient now

File: lib/services/sso_url_provider.js (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply lib/services/sso_url_provider.js]: After refactoring httpClient it is possible

**Review by sfc-gh-dprzybysz — APPROVED** [review]

**Pull request merged**